### DISCUSSION TITLE
Allow the rich-text editor to toggle raw markdown / mdx

### DISCUSSION THREAD

**Discussion by perkinsjr** [tinacms/tinacms#2571] [Ideas]
A user in the Discord brought up the idea of being able to toggle to view raw MDX editor.  What would be neat is if the editor itself could toggle between allowing a user to write in raw mdx vs rich text.

> We are testing out Tina for first time today, I'm super happy with it so far.

> Basically the situation is, we are ultimately writing in MDX so we want to leverage templates/custom components.  However we have reason to believe others on our team aren't in love with rich text editors and want to provide writing in markdown/mdx as an option for the editor.

> From my brief testing, if we use the markdown plugin for the editor we cannot use templates (I assume we can still write out raw mdx and it will be fine) we just dont get the benefits that the rich text editor provides in regards to templates.

> What would be neat is if the editor itself could toggle between allowing a user to write in raw mdx vs rich text.

I think is something we could offer as we offer it as part of the markdown editor today.

**Comment by jeffsee55** [comment]
This is not quite as simple as it might seem. My first thought is that implementing this could follow 2 paths:

1. `parse/stringify` the raw values to/from the editor
2. Make the "raw" format _feel_ like it's raw, but still be the formatted MDX AST

## Scenario 1

The main issue with this is that the `rich-text` editor needs a properly formatted MDX AST in order to work. To do this, we actually send the **parsed** version of MDX to and from the GraphQL API, so it's never treated as a string client-side. If we support going into raw mode it will mean we need to stringify and parse this stuff client-side. The real MDX implementation that has gained popularity is a _compilation_ tool, along with other Remark plugins, it assumes a Node environment is where all of this code is being run, though much of it will _probably_ work in the browser, I'm sure there will be a few issues.

---

Another issue is that this will likely lead to users running into more of the "Gotchas" section highlighted [here](https://github.com/tinacms/product-archive/discussions/26)
  - No support for import/export statements
  - Evaluating editor-supplied Javascript is a nogo. We should probably throw an error when we come across a non-literal expression.
    - We can support JSX expressions, but they must be [Literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types#literals).

I don't think there's much we can do here, if we run into people who want to be able to do more for the JavaScript-y things in MDX I'd love to learn more about their use case. From my perspective there's no scenario where code like this would/should work:

```md
---
---
// where does this come from? when does it get compiled?
import { MyComponent } from './components'

<MyComponent greeting="Hello, world!" />
```

## Scenario 2

> EDIT: eh, thinking more about this - I don't think it'll work, we'd need to bring a parser in for the JSX elements

I'll need to think a bit more about this, but it _may_ be possible, and if so would be much easier to implement than scenario 1.

We do things like automatically formatting `#### Hello` into this:
```js
{
  type: "h4",
  children: [{text: "Hello"}]
}
```

Perhaps we can skip the auto-formatting, and _then_ only do it when we emit the `onChange` event. So the editor would still think the value for `#### Hello` is:

```js
{
  type: "p",
  children: [{text: "#### Hello"}]
}
```
And then when we call the field's `onChange`, we do the formatting on the way out, so the `TinaMarkdown` component would still get the `h4` element as expected.

**Comment by llibdude** [comment]
I would still like this one day. I just realized where I saw this behavior when I was doing some initial research into git-based CMS's so I figure i'll update the ticket.  I found it on NetliflyCMS
![image](https://user-images.githubusercontent.com/5255924/162331130-17bb939b-baae-4374-a967-19e793d4b2b5.png)
